packages/block-editor-tools/src/hooks/use-option/index.ts

```diff
@@ -0,0 +1,72 @@
+import { Settings } from '@wordpress/core-data';
+import { useSelect, useDispatch } from '@wordpress/data';
+
+type ExtendedSettings<T> = (Settings & Record<string, T>);
+type RootSiteEntityRecordSelector<T> = (kind:string, name:string) => (
+  undefined | ExtendedSettings<T>
+);
+type RootSiteEntityRecordEditsSelector<T> = (kind:string, name:string) => (
+  undefined | Partial<ExtendedSettings<T>>
+);
+type UseOptionReturn<T> = {
+  value: T | undefined;
+  isEdited: boolean;
+  isSaving: boolean;
+  onChange: (newValue: T) => void;
+  onSave: () => Promise<void>;
+};
+type InferOptionType<K extends keyof any> = K extends keyof Settings
+  ? UseOptionReturn<Settings[K]>
+  : never;
+
+/**
+ * Custom hook for managing an option value in the WordPress block editor.
+ *
+ * @template T - The type of the option value. If the option is a built-in WordPress option, this
+ *               type will be inferred from the option key. If the option is a custom option, this
+ *               type must be provided.
+ */
+function useOption<K extends keyof Settings>(key: K): InferOptionType<K>;
+// eslint-disable-next-line no-redeclare
+function useOption<T>(key: string): UseOptionReturn<T>;
+// eslint-disable-next-line no-redeclare
+function useOption<K extends keyof Settings | string, T>(
+  key: K,
+): InferOptionType<K> | UseOptionReturn<T> {
+  const selectors = useSelect((select) => {
+    const { getEditedEntityRecord, getEntityRecordEdits, isSavingEntityRecord }: {
+      getEditedEntityRecord: RootSiteEntityRecordSelector<T>,
+      getEntityRecordEdits: RootSiteEntityRecordEditsSelector<T>,
+      isSavingEntityRecord: (kind:string, name:string) => boolean,
+    } = select('core');
```

Comment:
This has long been a frustration of mine with using `select` - it doesn't tell you what it's returning, so I've ended up suppressing TypeScript on lines where we're getting something from `select`. I like this approach, though - we're still assuming `select` will return what we think it will but we're at least describing what we expect it to return.

Reply:
Yeah, it feels really dirty to me to do this, but I couldn't figure out a better way.

side note, I spent _quite_ a while trying to figure out some better way, but it wasn't a total waste: it was in looking for a way to get types for these selectors that I stumbled upon `Settings`, which allowed me to do the inference juggling above. Before that, I only had `useOption<T>(...)`. Small win...